In [1]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt

np.random.seed(130298)

In [3]:
data = pd.read_csv('/Users/Nicole/OneDrive - Politecnico di Milano/Corsi/SL_health_care/project_heart_failure/example.csv')
data.set_index('inpatient.number', inplace = True)

## Verify if the variable with unbalance category are unbalance also in the outcome

Take the name of all the categorical and binary variables.

In [6]:
categorical_columns = ['visit.times','DestinationDischarge','admission.ward','admission.way','occupation','discharge.department',
                       'gender', 'type.of.heart.failure', 'NYHA.cardiac.function.classification', 'Killip.grade',
                       'consciousness','eye.opening','verbal.response', 'oxygen.inhalation', 'outcome.during.hospitalization',
                       'GCS','ageCat']

prova = categorical_columns.copy()
binary_columns = ['myocardial.infarction', 'congestive.heart.failure', 'peripheral.vascular.disease',
                  'cerebrovascular.disease', 'dementia', 'Chronic.obstructive.pulmonary.disease',
                  'peptic.ulcer.disease', 'diabetes', 'moderate.to.severe.chronic.kidney.disease',
                  'solid.tumor', 'liver.disease', 'type.II.respiratory.failure', 'death.within.28.days', 're.admission.within.28.days', 'death.within.3.months',
                  're.admission.within.3.months', 'death.within.6.months', 're.admission.within.6.months',
                  'return.to.emergency.department.within.6.months','respiratory.support.']
# NB 'respiratory.support.' is in the binary variables because we will rebin it in 0/1

prova.extend(binary_columns)

For each of the previous variables, verify if one category of the variables have more than 80% of the total values of the variables.

In [7]:
col_to_exclude=[]
for col_name in prova:
    max_value = (data[col_name].value_counts()/data.shape[0]).max()
    if max_value>0.8 and col_name!= "death.within.28.days" and col_name!= "re.admission.within.28.days" and col_name!= "death.within.3.months"and col_name!= "death.within.6.months":
        print(col_name,': ',max_value)
        col_to_exclude.append(col_name)

visit.times :  0.9262948207171314
occupation :  0.8316733067729084
discharge.department :  0.8481075697211156
consciousness :  0.9830677290836654
eye.opening :  0.9790836653386454
verbal.response :  0.9780876494023905
oxygen.inhalation :  0.9452191235059761
outcome.during.hospitalization :  0.9412350597609562
GCS :  0.9716135458167331
myocardial.infarction :  0.9287848605577689
congestive.heart.failure :  0.9322709163346613
peripheral.vascular.disease :  0.9497011952191236
cerebrovascular.disease :  0.9252988047808764
dementia :  0.9427290836653387
Chronic.obstructive.pulmonary.disease :  0.8839641434262948
peptic.ulcer.disease :  0.976593625498008
solid.tumor :  0.9805776892430279
liver.disease :  0.9576693227091634
type.II.respiratory.failure :  0.9432270916334662
respiratory.support. :  0.9790836653386454


Verify the distribution of 0s and 1s in the outcome variable: re.admission.within.6.months

In [13]:
outcome_name="re.admission.within.6.months"
number_1=len(data[data[outcome_name]==1])
number_0=len(data[data[outcome_name]==0])
print(number_1/(len(data)))

0.3849601593625498


The percentage of the re admission is 38.49%

Now, for each of the variables with more of 80% of unbalance, we verify if the distribution of the outcome is balance for the outcome. 

## visit.times


In [153]:
column_name="visit.times"
percentage=data[column_name].value_counts()/data.shape[0]
percentage

1    0.926295
2    0.059761
3    0.009960
4    0.002988
5    0.000996
Name: visit.times, dtype: float64

In [154]:
prova=data[data[column_name]==percentage.index[0]]
number_1=prova[prova[outcome_name]==1].shape[0]
number_0=prova[prova[outcome_name]==0].shape[0]
print(number_1/len(prova))

0.38064516129032255


Taking the row where visit.time=1 (the category with 92% of values), there are 38% of 1 in the outcome.

## occupation

In [155]:
column_name="occupation"
percentage=data[column_name].value_counts()/data.shape[0]
percentage

UrbanResident    0.831673
farmer           0.098606
Others           0.044323
worker           0.008466
Officer          0.003486
Name: occupation, dtype: float64

In [156]:
prova=data[data[column_name]==percentage.index[0]]
number_1=prova[prova[outcome_name]==1].shape[0]
number_0=prova[prova[outcome_name]==0].shape[0]
print(number_1/len(prova))

0.39820359281437123


Taking the row where occupation=UrbanResident (the category with 83% of values), there are 39.8% of 1 in the outcome.

## discharge.department

In [157]:
column_name="discharge.department"
percentage=data[column_name].value_counts()/data.shape[0]
percentage

Cardiology     0.848108
GeneralWard    0.120020
Others         0.025896
ICU            0.005976
Name: discharge.department, dtype: float64

In [158]:
prova=data[data[column_name]==percentage.index[0]]
number_1=prova[prova[outcome_name]==1].shape[0]
number_0=prova[prova[outcome_name]==0].shape[0]
print(number_1/len(prova))

0.39048737522019966


Taking the row where column_name="discharge.department" (the category with 84% of values), there are 39% of 1 in the outcome.

## eye.opening

In [12]:
column_name="eye.opening"
percentage=data[column_name].value_counts()/data.shape[0]
percentage

4    0.979084
3    0.012450
1    0.006972
2    0.001494
Name: eye.opening, dtype: float64

In [17]:
prova=data[data[column_name]==percentage.index[0]]
number_1=prova[prova[outcome_name]==1].shape[0]
number_0=prova[prova[outcome_name]==0].shape[0]
print(number_1/len(prova))

0.387589013224822


Taking the row where column_name="eye.opening" (the category with 97.9% of values), there are 38.7% of 1 in the outcome.

## verbal.response

In [161]:
column_name="verbal.response"
percentage=data[column_name].value_counts()/data.shape[0]
percentage

5    0.978088
3    0.008964
1    0.006972
2    0.003984
4    0.001992
Name: verbal.response, dtype: float64

In [162]:
prova=data[data[column_name]==percentage.index[0]]
number_1=prova[prova[outcome_name]==1].shape[0]
number_0=prova[prova[outcome_name]==0].shape[0]
print(number_1/len(prova))

0.3874745417515275


Taking the row where column_name="verbal.response" (the category with 97.8% of values), there are 38.7% of 1 in the outcome.

## oxygen.inhalation


In [163]:
column_name="oxygen.inhalation"
percentage=data[column_name].value_counts()/data.shape[0]
percentage

OxygenTherapy    0.945219
AmbientAir       0.054781
Name: oxygen.inhalation, dtype: float64

In [164]:
prova=data[data[column_name]==percentage.index[0]]
number_1=prova[prova[outcome_name]==1].shape[0]
number_0=prova[prova[outcome_name]==0].shape[0]
print(number_1/len(prova))

0.38250790305584825


Taking the row where column_name="oxygen.inhalation" (the category with 94.5% of values), there are 38.2% of 1 in the outcome.

## outcome.during.hospitalization

In [165]:
column_name="outcome.during.hospitalization"
percentage=data[column_name].value_counts()/data.shape[0]
percentage

Alive                    0.941235
DischargeAgainstOrder    0.053287
Dead                     0.005478
Name: outcome.during.hospitalization, dtype: float64

In [166]:
prova=data[data[column_name]==percentage.index[0]]
number_1=prova[prova[outcome_name]==1].shape[0]
number_0=prova[prova[outcome_name]==0].shape[0]
print(number_1/len(prova))

0.3941798941798942


Taking the row where column_name="outcome.during.hospitalization" (the category with 94.1% of values), there are 39.4% of 1 in the outcome.

## GCS

In [167]:
column_name="GCS"
percentage=data[column_name].value_counts()/data.shape[0]
percentage

15    0.971614
11    0.009462
3     0.006474
10    0.003486
14    0.003486
7     0.001992
12    0.001494
13    0.000996
6     0.000498
4     0.000498
Name: GCS, dtype: float64

In [168]:
prova=data[data[column_name]==percentage.index[0]]
number_1=prova[prova[outcome_name]==1].shape[0]
number_0=prova[prova[outcome_name]==0].shape[0]
print(number_1/len(prova))

0.3885187083546899


Taking the row where column_name="GCS" (the category with 97.1% of values), there are 38.8% of 1 in the outcome.

## myocardial.infarction

In [169]:
column_name="myocardial.infarction"
percentage=data[column_name].value_counts()/data.shape[0]
percentage

0    0.928785
1    0.071215
Name: myocardial.infarction, dtype: float64

In [170]:
prova=data[data[column_name]==percentage.index[0]]
number_1=prova[prova[outcome_name]==1].shape[0]
number_0=prova[prova[outcome_name]==0].shape[0]
print(number_1/len(prova))

0.3871313672922252


Taking the row where column_name="myocardial.infarction" (the category with 92.8% of values), there are 38.7% of 1 in the outcome.

## congestive.heart.failure

In [171]:
column_name="congestive.heart.failure"
percentage=data[column_name].value_counts()/data.shape[0]
percentage

1    0.932271
0    0.067729
Name: congestive.heart.failure, dtype: float64

In [172]:
prova=data[data[column_name]==percentage.index[0]]
number_1=prova[prova[outcome_name]==1].shape[0]
number_0=prova[prova[outcome_name]==0].shape[0]
print(number_1/len(prova))

0.38247863247863245


Taking the row where column_name="congestive.heart.failure" (the category with 93.2% of values), there are 38.2% of 1 in the outcome.

## peripheral.vascular.disease

In [173]:
column_name="peripheral.vascular.disease"
percentage=data[column_name].value_counts()/data.shape[0]
percentage

0    0.949701
1    0.050299
Name: peripheral.vascular.disease, dtype: float64

In [174]:
prova=data[data[column_name]==percentage.index[0]]
number_1=prova[prova[outcome_name]==1].shape[0]
number_0=prova[prova[outcome_name]==0].shape[0]
print(number_1/len(prova))

0.38437336130047195


Taking the row where column_name="peripheral.vascular.disease" (the category with 94.9% of values), there are 38.% of 1 in the outcome.

## cerebrovascular.disease

In [175]:
column_name="cerebrovascular.disease"
percentage=data[column_name].value_counts()/data.shape[0]
percentage

0    0.925299
1    0.074701
Name: cerebrovascular.disease, dtype: float64

In [19]:
prova=data[data[column_name]==percentage.index[0]]
number_1=prova[prova[outcome_name]==1].shape[0]
number_0=prova[prova[outcome_name]==0].shape[0]
print(number_1/len(prova))

0.387589013224822


Taking the row where column_name="cerebrovascular.disease" (the category with 92.5% of values), there are 38.4% of 1 in the outcome.

## dementia

In [20]:
column_name="dementia"
percentage=data[column_name].value_counts()/data.shape[0]
percentage

0    0.942729
1    0.057271
Name: dementia, dtype: float64

In [21]:
prova=data[data[column_name]==percentage.index[0]]
number_1=prova[prova[outcome_name]==1].shape[0]
number_0=prova[prova[outcome_name]==0].shape[0]
print(number_1/len(prova))

0.37770734284204965


Taking the row where column_name="dementia" (the category with 94.2% of values), there are 37.7% of 1 in the outcome.

## Chronic.obstructive.pulmonary.disease

In [179]:
column_name="Chronic.obstructive.pulmonary.disease"
percentage=data[column_name].value_counts()/data.shape[0]
percentage

0    0.883964
1    0.116036
Name: Chronic.obstructive.pulmonary.disease, dtype: float64

In [180]:
prova=data[data[column_name]==percentage.index[0]]
number_1=prova[prova[outcome_name]==1].shape[0]
number_0=prova[prova[outcome_name]==0].shape[0]
print(number_1/len(prova))

0.38535211267605635


Taking the row where column_name="Chronic.obstructive.pulmonary.disease" (the category with 88.3% of values), there are 38.5% of 1 in the outcome.

## peptic.ulcer.disease

In [181]:
column_name="peptic.ulcer.disease"
percentage=data[column_name].value_counts()/data.shape[0]
percentage

0.0    0.976594
1.0    0.022410
Name: peptic.ulcer.disease, dtype: float64

In [182]:
prova=data[data[column_name]==percentage.index[0]]
number_1=prova[prova[outcome_name]==1].shape[0]
number_0=prova[prova[outcome_name]==0].shape[0]
print(number_1/len(prova))

0.38245792962774094


Taking the row where column_name="peptic.ulcer.disease" (the category with 97.6% of values), there are 38.2% of 1 in the outcome.

## solid.tumor

In [183]:
column_name="solid.tumor"
percentage=data[column_name].value_counts()/data.shape[0]
percentage

0    0.980578
1    0.019422
Name: solid.tumor, dtype: float64

In [184]:
prova=data[data[column_name]==percentage.index[0]]
number_1=prova[prova[outcome_name]==1].shape[0]
number_0=prova[prova[outcome_name]==0].shape[0]
print(number_1/len(prova))

0.3854748603351955


Taking the row where column_name="solid.tumor" (the category with 98% of values), there are 38.5% of 1 in the outcome.

## liver.disease

In [22]:
column_name="liver.disease"
percentage=data[column_name].value_counts()/data.shape[0]
percentage

0.0    0.957669
1.0    0.041833
Name: liver.disease, dtype: float64

In [23]:
prova=data[data[column_name]==percentage.index[0]]
number_1=prova[prova[outcome_name]==1].shape[0]
number_0=prova[prova[outcome_name]==0].shape[0]
print(number_1/len(prova))

0.3858554342173687


Taking the row where column_name="liver.disease" (the category with 95.7% of values), there are 38.5% of 1 in the outcome.

## type.II.respiratory.failure

In [25]:
column_name="type.II.respiratory.failure"
percentage=data[column_name].value_counts()/data.shape[0]
percentage

NonTypeII    0.943227
TypeII       0.056773
Name: type.II.respiratory.failure, dtype: float64

In [26]:
prova=data[data[column_name]==percentage.index[0]]
number_1=prova[prova[outcome_name]==1].shape[0]
number_0=prova[prova[outcome_name]==0].shape[0]
print(number_1/len(prova))

0.38753959873284055


Taking the row where column_name="type.II.respiratory.failure" (the category with 94.3% of values), there are 38.7% of 1 in the outcome.

## respiratory.support.

In [28]:
column_name="respiratory.support."
percentage=data[column_name].value_counts()/data.shape[0]
percentage

None    0.979084
IMV     0.012450
NIMV    0.008466
Name: respiratory.support., dtype: float64

In [32]:
prova=data[data[column_name]==percentage.index[0]]
number_1=prova[prova[outcome_name]==1].shape[0]
number_0=prova[prova[outcome_name]==0].shape[0]
print(number_1/len(prova))

0.38657171922685657


Taking the row where column_name="respiratory.support." (the category with 97.9% of values), there are 38.6% of 1 in the outcome.